components/tidb_query_normal_expr/src/builtin_time.rs

```diff
@@ -1,16 +1,27 @@
 // Copyright 2017 TiKV Project Authors. Licensed under Apache-2.0.
 
+use regex::Regex;
+
 use std::borrow::Cow;
 
 use crate::ScalarFunc;
 use tidb_query_datatype::codec::error::Error;
 use tidb_query_datatype::codec::mysql::time::extension::DateTimeExtension;
 use tidb_query_datatype::codec::mysql::time::weekmode::WeekMode;
-use tidb_query_datatype::codec::mysql::{Duration as MyDuration, Time, TimeType};
+use tidb_query_datatype::codec::mysql::{Duration as MyDuration, Time, TimeType, MAX_FSP};
 use tidb_query_datatype::codec::Datum;
 use tidb_query_datatype::expr::SqlMode;
 use tidb_query_datatype::expr::{EvalContext, Result};
 
+// is_duration returns a boolean indicating whether the str matches the format of duration.
+// See https://dev.mysql.com/doc/refman/5.7/en/time.html
+fn is_duration(s: &str) -> bool {
+    let re =
+        Regex::new(r"^\s*[-]?(((\d{1,2}\s+)?0*\d{0,3}(:0*\d{1,2}){0,2})|(\d{1,7}))?(\.\d*)?\s*$")
+            .unwrap();
+    re.is_match(s)
```

Comment:
How's the benchmark of this regex?

Reply:
Just copied the regex from TiDB. I'll add the related benchmark test.

Reply:
Benchmark test simply added. On my machine (2 GHz Quad-Core Intel Core i5 & 16 GB 3733 MHz LPDDR4) avg time ~0.45 ms/iter.